### PR TITLE
Add experimentation data-layer

### PR DESCRIPTION
**Pull request by withinboredom**
#### Changes proposed in this Pull Request

* Adds the experiment data-layer from #38755
* Depends on #38907 to be merged

Todo:

- [x] add tests

#### Testing instructions

* run `npm test` to run the tests

### CODE REVIEW

File: client/state/data-layer/wpcom/experiments/index.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+/**
+ * Transform the result from the API into the action we can use
+ *
+ * @param data The result from the API
+ * @returns {{variations: object, nextRefresh: number}} The transformed result
+ */
+const transformApiRequest = data => ( {
+	variations: data.variations,
+	nextRefresh: data.ttl,
```
Comment (yanirs) [inline client/state/data-layer/wpcom/experiments/index.js]: Shouldn't this be a function of the current time? Something like `now + ttl`?

Comment (withinboredom) [reply client/state/data-layer/wpcom/experiments/index.js]: Yep, good point!

File: client/state/data-layer/wpcom/experiments/index.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import { assignToExperiments } from 'state/experiments/actions';
+import config from 'config';
+
+/**
+ * Transform the result from the API into the action we can use
+ *
+ * @param data The result from the API
+ * @returns {{variations: object, nextRefresh: number}} The transformed result
+ */
+const transformApiRequest = data => ( {
```
Comment (jsnmoon) [inline client/state/data-layer/wpcom/experiments/index.js]: Do we want to type this function?

**Review by jsnmoon — COMMENTED** [review]
The code looks good. 

When will we be able to test this end-to-end?

File: client/state/data-layer/wpcom/experiments/index.js (hunk 2 above)
Comment (withinboredom) [reply client/state/data-layer/wpcom/experiments/index.js]: I wanted to use typescript here, but you run into being unable to type a bunch of things properly.

**Comment by withinboredom** [comment]
> When will we be able to test this end-to-end?

After these two PRs are merged, we'll add the components. You'll still need a dummy API endpoint, but work there has also been started.

**Review by yanirs — APPROVED** [review]

File: client/state/data-layer/wpcom/experiments/index.js (hunk 2 above)
Comment (jsnmoon) [reply client/state/data-layer/wpcom/experiments/index.js]: Alternatively, you could spin the transform function into a separate typescript file and import it here. Just food for though.

**Review by dzver — APPROVED** [review]
LGTM

**Pull request merged**